File: Share extension dismiss/objc/MyShareExtensionDismissControlRecognizer.m
//
//  MyShareExtensionDismissControlRecognizer.m
//  MyShareExtension
//
//  Created by Huan Lin on 3/28/25.
//

#import "MyShareExtensionDismissControlRecognizer.h"

@interface MyShareExtensionDismissControlRecognizer()<UIGestureRecognizerDelegate>

@end

@implementation MyShareExtensionDismissControlRecognizer {
  CGPoint _beganLocation;
  UIGestureRecognizer *_dismissRecognizer;
  MyShareExtensionSwipeDismissStrategy _dismissStrategy;
}

- (instancetype)initWithDismissStrategy:(MyShareExtensionSwipeDismissStrategy)strategy {
  self = [super init];
  if (self) {
    _dismissStrategy = strategy;
    self.delegate = self;
  }
  return self;
}

- (BOOL)isBackgroundDismissRecognizer: (UIGestureRecognizer *)gc {
  return [gc.name hasSuffix:@"UISheetInteractionBackgroundDismissRecognizer"];
}

- (BOOL)shouldRequireFailureOfGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {

  if (!_dismissRecognizer && [self isBackgroundDismissRecognizer:otherGestureRecognizer]) {
    _dismissRecognizer = otherGestureRecognizer;
    
    switch (_dismissStrategy) {
      case MyShareExtensionSwipeDismissStrategyTopRegion:
        [self setDismissRecognizerEnabledWithLocation:_beganLocation];
        break;
      case MyShareExtensionSwipeDismissStrategyProhibited:
        _dismissRecognizer.enabled = NO;
        break;
    }
  }
  
  return NO;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  _beganLocation = [touches.anyObject locationInView:self.view];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  self.state = UIGestureRecognizerStateFailed;
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  self.state = UIGestureRecognizerStateFailed;
}

- (void)setDismissRecognizerEnabledWithLocation: (CGPoint)location {
  _dismissRecognizer.enabled = location.y <= 50;
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
  switch (_dismissStrategy) {
    case MyShareExtensionSwipeDismissStrategyTopRegion:
      [self setDismissRecognizerEnabledWithLocation:[touch locationInView:touch.view]];
      break;
    case MyShareExtensionSwipeDismissStrategyProhibited:
      break;
  }
  return YES;
}

@end

File: Share extension dismiss/swift/MyShareExtensionDismissControlRecognizer.swift
//
//  MyShareExtensionDismissControlRecognizer.swift
//  MyShareExtension
//
//  Created by Huan Lin on 3/31/25.
//

import UIKit

final class MyShareExtensionDismissControlRecognizer: UIGestureRecognizer {
  
  enum DismissStrategy {
    case topRegion
    case prohibited
  }
  
  private let strategy: DismissStrategy
  
  private var beganLocation: CGPoint = .zero
  private var dismissRecognizer: UIGestureRecognizer? = nil
  
  init(strategy: DismissStrategy) {
    self.strategy = strategy
    super.init(target: nil, action: nil)
    
    delegate = self
  }
  
  private func setDismissRecognizerEnabled(location: CGPoint) {
    dismissRecognizer?.isEnabled = location.y <= 50
  }
  
  override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent) {
    guard let touch = touches.first else { return }
    beganLocation = touch.location(in: view)
  }
  
  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent) {
    state = .failed
  }
  
  override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent) {
    state = .cancelled
  }
}

extension MyShareExtensionDismissControlRecognizer: UIGestureRecognizerDelegate {
  
  func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRequireFailureOf otherGestureRecognizer: UIGestureRecognizer) -> Bool {
  
    if (dismissRecognizer == nil && otherGestureRecognizer.isBackgroundDismissRecognizer) {
      dismissRecognizer = otherGestureRecognizer
      
      switch strategy {
      case .topRegion:
        setDismissRecognizerEnabled(location: beganLocation)
      case .prohibited:
        dismissRecognizer?.isEnabled = false
      }
    }
    return false
  }
  
  func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
    switch strategy {
    case .topRegion:
      setDismissRecognizerEnabled(location: touch.location(in: touch.view))
    case .prohibited:
      break
    }
    return true
  }
}

fileprivate extension UIGestureRecognizer {
  var isBackgroundDismissRecognizer: Bool {
    return name?.hasSuffix("UISheetInteractionBackgroundDismissRecognizer") ?? false
  }
}
